pmlc/dialect/pxa/transforms/autotile.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <limits>
+#include <memory>
+#include <vector>
+
+#include "mlir/Analysis/AffineStructures.h"
+#include "mlir/Pass/Pass.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/dialect/pxa/transforms/tile.h"
+
+namespace pmlc::dialect::pxa {
+
+// A tile size generator is a functor from range -> list of sizes
+
+// Produces all powers of 2 <= range
+std::vector<int64_t> PowerOfTwoGenerator(int64_t range);
+
+// Produces all divisors of range (including the full range)
+std::vector<int64_t> EvenTilingGenerator(int64_t range);
+
+// Produces only 'range' itself
+inline std::vector<int64_t> ExactRangeGenerator(int64_t range) {
+  return {range};
+}
+
+// Produces only a specific value
+class FixedTileSizeGenerator {
+public:
+  explicit FixedTileSizeGenerator(int64_t val) : val(val) {}
+  std::vector<int64_t> operator()(int64_t range) const { return {val}; }
+
+private:
+  int64_t val;
+};
+
+// Producs the union of multiple generators
+template <typename Head, typename... Rest>
+class UnionGenerator {
+public:
+  UnionGenerator(const Head &head, const Rest &... rest)
```

Comment:
You might want to use `&&` instead of `const &` to be able to have perfect forwarding...

Reply:
Also use `std::forward` when passing `rest`